doc/memos/rdm0002.md

```diff
@@ -0,0 +1,2226 @@
+`IEEE802154_RADIO_INDICATION_RX_DONE` event might post an event to an event
+queue in order to fetch the packet).
+
+The full list of events and implications are defined in the Interface
+Definition section.
+
+### 4.3.3 Device Driver
+
+The Device Driver implements the hardware-dependent part of the IEEE802.15.4
+Radio HAL by wrapping the `radio_ops` interface around the device specific
+code, which grants access to all device operations.
+
+The Device Driver additionally provides a mechanism to expose the ISR of
+radios that require the Bottom-Half processor.
+
+The function set of the Device Driver can include device specific features that
+are not exposed by the Radio HAL API (e.g., Smart Listening with AT86RF2xx
+radios).
+
+# 5 Implementation Details
+## 5.1 Initialization of device drivers
+In order to implement the 802.15.4 abstraction on top of a device driver, it
+is required an initialization procedure that performs the following tasks:
+
+1. Set up IRQ callback
+2. Reset the device
+3. Confirm connectivity and perform self tests
+4. Bring device into a low power state
+5. Set up IRQs and disable them to use less power.
+
+The `radio_ops` interface provides an "on" function that turns on the device
+and enables interrupts. It is expected that the upper layer will call this
+function to enable the radio device, if the initialization procedure succeeded.
+
+## 5.2 Abstract State Machine
+
+The Radio HAL defines an Abstract State Machine. In order to ensure a uniform
+behavior in all devices, all Radio HAL device drivers should be implemented
+against this. Transient states (e.g pending requests) are not included in this
+diagram. However, the upper layer MUST NOT trigger another request if there is
+already a pending one.
+
+```
+                                 +---------+
+                                 |         |
+                                 |   OFF   |<------ Any state
+                                 |         |   OFF
+                                 +---------+
+                                   |
+                            ON     |
+                                   |
+                                   v
+                                 +---------+
+                        +--------|         |--------+
+                        |        | TRX_OFF |<----+  |
+                        | +----->|         |     |  |
+                        | |      +---------+     |  |
+         SET_TRX_STATE  | |                      |  | SET_TRX_STATE
+                        | |                      |  |
+                        v |                      |  v
+                    +---------+              +---------+
+                    |         |------------->|         |
+                    |   IDLE  |              |   RX    |
+                    |         |<-------------|         |
+                    +---------+              +---------+
+                               SET_TRX_STATE
```

Comment:
Getting back to this, is there not an implicit TX state when the device is busy sending the actual frame?